Reword and add uniformity to Standard Library documentation (#4464)
edgedb/edgedb: docs/stdlib/abstract.rst

Abstract Types
==============

Abstract types are used to describe a polymorphic, otherwise known as

> generic function to be applicable over a broad range of types.

Reviewer: Grammar nerd here. Since we're talking about "types" (plural), I think we should say that they describe "polymorphic function**s**" (also plural) rather than "a polymorphic function" (singular)

Also, I would add another comma after "function" to fully set off that phrase from the rest of the sentence.

Suggested rewrite:
> Abstract types are used to describe polymorphic functions, otherwise known as "generic functions," which can return a broad range of types.

> Abstract types are used to describe a polymorphic, otherwise known as
> generic function to be applicable over a broad range of types.

Reviewer: New suggestion taking into account @vpetrovykh's feedback:
```suggestion
Abstract types are used to describe polymorphic functions, otherwise known as
"generic functions," which can be called on a broad range of value types.
```

    :index: any anytype

>     Represents a generic type.

Reviewer: In these instances where you've changed the description to read "_represents_" a particular type, is that accurate, or _are_ these that type? Use of the word "represents" makes me think these are some sort of stand-in for an actual type, but I believe they _are_ these types.

Author: "Represents" is added just to inform users that this is the actual representation of the type's definition and construct in the documentation

Reviewer: I don't think this should define the documentation though. We're documenting the type, so the description should describe the type, not the documentation of the type.

    This type is used as a placeholder in cases where there isn't a specific

>     set of type requirements, or are not necessary, such as defining

Reviewer: The second half of this is a bit difficult to read. Suggestion:
> ... where there isn't a specific set of type requirements, or where strict typing is not necessary, such as ...

>     This type is used as a placeholder in cases where there isn't a specific
>     set of type requirements, or are not necessary, such as defining
>     polymorphic parameters in functions and operators.

Reviewer: Suggestion considering all the feedback so far:
```suggestion
    This type is used as a placeholder in cases where there aren't specific
    type requirements such as when defining polymorphic parameters in functions
    and operators.
```

    :index: any anytype anytuple

    Represents a generic tuple.

    Similar to :eql:type:`anytype`, this type is used to denote a generic

>     tuple without going into detail of which components there are. This is

Reviewer: Suggested change for clarity:
> ... to denote a generic tuple without typing each of its constituents. This is useful ...

>     tuple without going into detail of which components there are. This is

Reviewer: ```suggestion
    tuple without detailing its constituents. This is useful when defining
```

>     useful when defining polymorphic parameters in functions and operators.

Reviewer: ```suggestion
    polymorphic parameters in functions and operators.
```

Abstract Numeric Types
======================

> Abstract numeric types are used when wanting to extend off of

Author: Poor wording, should have been kept as the former. Thanks for catching this Devon

    :index: any anytype

    Represents an abstract base scalar type for
    :eql:type:`anyint`, :eql:type:`anyfloat` and :eql:type:`decimal`.

Abstract Range Types
====================

Like abstract types, you may also specify :ref:`ranges <ref_std_range>`.

> These scalar types are distinguished by the following:

Reviewer 2: Hm, I've trouble parsing the new version. `Like abstract types, you may also...` is making me think that the types we discuss in this section are not "abstract" but they literally are.

`These scalar types are distinguished by the following:` -- "distinguished" works weirdly here. Also "these scalar types" looks misplaced, this section and the paragraph talks about abstract types (well, they're also scalar, but that's not the point here!)

All in all, I agree that the current wording can be improved, but I don't quite like the new suggested one.

Author: @1st1 I think changing L101 to `With abstract types, [...]` may be best to eliminate confusion from wording. Let me know what you think.

Good catch on L102, this is misplaced writing, I modified the original wording which could mean this may have been here for a bit? I'll think on this, maybe making a generic reference to this with `Types can be distinguished by the following:` is more appropriate -- will think on this.

    :index: any anytype

    Represents an abstract base scalar type for
    :eql:type:`anyint`, :eql:type:`anyfloat` and :eql:type:`decimal` types.

Abstract Range Types
====================

> Like abstract numeric types, :ref:`ranges <ref_std_range>` can also be

Reviewer: ```suggestion
These types serve as the base types for all :ref:`ranges <ref_std_range>`.
```

> specified.

Reviewer: ```suggestion
```

.. eql:type:: std::anypoint

>     :index: any anypoint anyrange point

Reviewer: Recommended not to add these additional search terms to the range types since a user searching for those new terms is not likely to be looking for these abstract types. Same recommendation on line 123.

Reviewer: ```suggestion
    :index: any anypoint anyrange
```

>     :index: any anydiscrete anyrange

Reviewer 2: ```suggestion
    :index: any anydiscrete anyrange discrete
```
